platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/io/ParallelOutputStream.kt

```diff
@@ -0,0 +1,248 @@
+            }
+        }
+        return QueuedOutputStream(buffers, ready) {
+            writer.join()
+        }
+    }
+}
+
+
+/**
+ * An [OutputStream] implementation that writes to buffers taken from a [ByteBufferPool]
+ * and posts them to the given [ready] queue when they are full.
+ */
+private
+class QueuedOutputStream(
+    private val buffers: ByteBufferPool,
+    private val ready: Queue<ByteBuffer>,
+    private val onClose: () -> Unit,
+) : OutputStream() {
+
+    private
+    var buffer = buffers.take()
+
+    override fun write(b: ByteArray, off: Int, len: Int) {
+        writeByteArray(b, off, len)
+    }
+
+    override fun write(b: Int) {
+        buffer.put(b.toByte())
+        maybeFlush()
+    }
+
+    override fun close() {
+        // send remaining data
+        if (buffer.position() > 0) {
+            sendBuffer()
+            takeNextBuffer()
+        }
+        // send a last empty buffer to signal the end
+        sendBuffer()
+        onClose()
+        buffers.rethrowFailureIfAny()
+        super.close()
+    }
+
+    private
+    tailrec fun writeByteArray(b: ByteArray, off: Int, len: Int) {
+        if (len <= 0) {
+            return
+        }
+        val remaining = buffer.remaining()
+        if (remaining > len) {
+            putByteArrayAndFlush(b, off, len)
+        } else {
+            putByteArrayAndFlush(b, off, remaining)
+            writeByteArray(b, off + remaining, len - remaining)
+        }
+    }
+
+    private
+    fun putByteArrayAndFlush(b: ByteArray, off: Int, len: Int) {
+        buffer.put(b, off, len)
+        maybeFlush()
+    }
+
+    private
+    fun maybeFlush() {
+        if (!buffer.hasRemaining()) {
+            sendBuffer()
+            takeNextBuffer()
+        }
+    }
+
+    private
+    fun sendBuffer() {
+        buffer.flip()
+        ready.offer(buffer)
+    }
+
+    private
+    fun takeNextBuffer() {
+        buffer = buffers.take()
+    }
+}
+
+
+/**
+ * Manages a pool of buffers of fixed [capacity][ByteBufferPool.bufferCapacity] allocated on-demand
+ * upto a [fixed maximum][ByteBufferPool.bufferCount].
+ */
+private
+class ByteBufferPool {
+
+    companion object {
+
+        /**
+         * How many bytes are transferred, at a time, from the producer to the writer thread.
+         *
+         * The smaller the number the more parallelism between producer and writer.
+         * The default is `32` for increased parallelism.
+         */
+        val bufferCapacity = System.getProperty("org.gradle.configuration-cache.internal.buffer-capacity", null)?.toInt()
+            ?: 32
+
+        /**
+         * Maximum number of buffers to be allocated.
+         *
+         * Determines the maximum memory working set: [bufferCount] * [bufferCapacity].
+         * The default maximum working set is `32MB`.
+         */
+        val bufferCount = System.getProperty("org.gradle.configuration-cache.internal.buffer-count", null)?.toInt()
```

Comment:
Is this right? 1 million buffer that are 32 bit long each? Normally, I'd expect the other way around.

Reply:
> Is this right? 1 million buffers that are 32 bit long each? Normally, I'd expect the other way around.

Yes, perhaps the term buffer is not the best choice here. A better analogy might be of a `packet` (like in UDP). These are used to quickly send the encoded information from the producer to the writer thread for compression >> encryption >> disk IO.

Reply:
@abstratt Please take another look, I changed the code to the `packet` analogy.

Reply:
> 32 bit long 
32 bytes